url43a0865bb0_kevinjdolan_darwin_tgz-pJ8-tui_DriverJ8/src/darwInvest/data/Ticker.java

```diff
@@ -0,0 +1,373 @@
+	 * @param which the index of the data field
+	 * @param data  the value of the data
+	 * @return		a number (preferably) [0,1]
+	 */
+	public double normalizeData(int which, double data) {
+		double diff = data - minData[which];
+		double range = maxData[which] - minData[which];
+		return diff / range;
+	}
+	
+	/**
+	 * Get a list of all news stories in a given time range
+	 * @param firstTime the beginning of the time range
+	 * @param lastTime	the end of the time range
+	 * @return			the list of news events
+	 */
+	public List<NewsEvent> getNews(long firstTime, long lastTime) {
+		Collection<List<NewsEvent>> collection = 
+			(news.subMap(firstTime, true, lastTime,true)).values();
+		List<NewsEvent> list = new ArrayList<NewsEvent>();
+		for(List<NewsEvent> entry : collection)
+			for(NewsEvent event : entry)
+				list.add(event);
+		return list;
+	}
+	
+	/**
+	 * @return the first time of this data set (considering only price information)
+	 */
+	public long getFirstTime() {
+		return data.firstKey();
+	}
+	
+	/**
+	 * @return the last time of this data set (considering only price information)
+	 */
+	public long getLastTime() {
+		return data.lastKey();
+	}
+	
+	/**
+	 * @return the first time of the news data set
+	 */
+	public long getFirstNews() {
+		return news.firstKey();
+	}
+	
+	/**
+	 * @return the last time of the news data set
+	 */
+	public long getLastNews() {
+		return news.lastKey();
+	}
+	
+	/**
+	 * Return all times between the given range
+	 * @param begin the beginning of the range
+	 * @param end   the end of the range
+	 * @return      an ordered set of time values
+	 */
+	public Set<Long> getTimesBetween(long begin, long end) {
+		return data.subMap(begin, end).keySet();
+	}
+	
+	/**
+	 * Get the lowest time greater than the input time
+	 * @param time the time to look for
+	 * @return     the ceiling time
+	 */
+	public long ceilingTime(long time) {
+		return data.ceilingKey(time);
+	}
+	
+	/**
+	 * Get the greatest time less than the input time
+	 * @param time the time to look for
+	 * @return     the floor time
+	 */
+	public long floorTime(long time) {
+		return data.floorKey(time);
+	}
+
+	public String lookupGooglecompanyid() throws IOException {
+		try {
+			URL google = new URL("http://www.google.com/finance?q=NASDAQ:" + symbol);
+			BufferedReader in = new BufferedReader(new InputStreamReader(google.openStream()));
+			Scanner s = new Scanner(in);
+			s.useDelimiter("[\n\r]+");
+			String companyid = null;
+			while(s.hasNext()) {
+				String next = s.next();
+				if(next.contains("var _companyId")) {
+					String extractor = next.substring(17);
+					companyid = extractor.substring(0,(extractor.indexOf(';')));
+					break;
+				}
+			}
+			if (s != null) {
+				s.close();
+			}
+			if (in != null) {
+				in.close();
+			}
+			this.googlecompanyid = companyid;
+			return companyid;
+		}
+		catch (Exception e) {
+			throw new IOException();
+		}
+	}
+
+	public String lookupCompanyname() throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 317-334 into a separate method.